File: web/lib/hedgeMath.ts
/**
 * Two-state hedge sizing (mirrors structuring/hedge_math.py + hedgingresearch notebooks).
 *
 * N_yes = notional × (r_no − r_yes) locks wealth across YES/NO resolution states.
 */

export interface HedgeCalibration {
  /** Signed equity return when the adverse outcome hits. */
  moveAdverse: number;
  /** Signed equity return when the favorable outcome hits. */
  moveFavorable: number;
  /** Implied P(YES) on the matched market question. */
  yesProbability: number;
  /** Side we buy to execute the hedge. */
  hedgeLeg: "YES" | "NO";
  /** Optional half-spread added to YES cost (0..1). */
  spread?: number;
}

export interface TwoStateSizing {
  retIfYes: number;
  retIfNo: number;
  pYes: number;
  yesCost: number;
  nYes: number;
  contracts: number;
  execSide: "YES" | "NO";
  entryPrice: number;
  premiumUsd: number;
  premiumSignedUsd: number;
  wealthGapUsd: number;
  hedgeQuality: number;
  approachNote: string;
}

export function yesReturnsFromLegs(
  moveAdverse: number,
  hedgeLeg: "YES" | "NO",
  moveFavorable = 0,
): { retIfYes: number; retIfNo: number } {
  if (hedgeLeg === "YES") {
    return { retIfYes: moveAdverse, retIfNo: moveFavorable };
  }
  return { retIfYes: moveFavorable, retIfNo: moveAdverse };
}

export function approachNote(
  retIfYes: number,
  retIfNo: number,
  pYes: number,
  nYes: number,
): string {
  const pct = (x: number) => `${x >= 0 ? "+" : ""}${Math.round(x * 100)}%`;
  const action = nYes >= 0 ? "buy YES" : "buy NO (short YES)";
  return `Two-state lock: N = notional × (${pct(retIfNo)} − ${pct(retIfYes)}) at P(YES)=${Math.round(pYes * 100)}%; ${action} equalizes wealth across outcomes.`;
}

export function sizeTwoState(
  notional: number,
  cal: HedgeCalibration,
): TwoStateSizing {
  const { retIfYes, retIfNo } = yesReturnsFromLegs(
    cal.moveAdverse,
    cal.hedgeLeg,
    cal.moveFavorable,
  );
  const spread = cal.spread ?? 0;
  const p = Math.min(Math.max(cal.yesProbability, 0.01), 0.99);
  const yesCost = Math.min(p + spread, 0.999);

  const nYes = notional * (retIfNo - retIfYes);

  let execSide: "YES" | "NO";
  let contracts: number;
  let entryPrice: number;

  if (nYes >= 0) {
    execSide = "YES";
    contracts = Math.abs(nYes);
    entryPrice = yesCost;
  } else {
    execSide = "NO";
    contracts = Math.abs(nYes);
    entryPrice = 1 - yesCost;
  }

  const premiumSignedUsd = nYes * yesCost;
  const premiumUsd = Math.abs(premiumSignedUsd);

  const wYes = notional * (1 + retIfYes) + nYes * (1 - yesCost);
  const wNo = notional * (1 + retIfNo) - nYes * yesCost;
  const wealthGapUsd = Math.abs(wNo - wYes);
  const spreadRet = Math.abs(retIfNo - retIfYes);
  const hedgeQuality =
    spreadRet > 0 ? Math.max(0, Math.min(1, 1 - wealthGapUsd / (notional * spreadRet))) : 1;

  return {
    retIfYes,
    retIfNo,
    pYes: p,
    yesCost,
    nYes,
    contracts: Math.round(contracts),
    execSide,
    entryPrice,
    premiumUsd,
    premiumSignedUsd,
    wealthGapUsd,
    hedgeQuality,
    approachNote: approachNote(retIfYes, retIfNo, p, nYes),
  };
}

/** Scale full two-state sizing by a hedge ratio in [0, 1]. */
export function scaleSizing(
  notional: number,
  s: TwoStateSizing,
  ratio: number,
): TwoStateSizing {
  const r = Math.max(0, Math.min(1, ratio));
  const nYes = s.nYes * r;
  const premiumSignedUsd = nYes * s.yesCost;
  const wYes = hedgedWealth(notional, s.retIfYes, nYes, s.yesCost, true);
  const wNo = hedgedWealth(notional, s.retIfNo, nYes, s.yesCost, false);
  const wealthGapUsd = Math.abs(wNo - wYes);
  const spreadRet = Math.abs(s.retIfNo - s.retIfYes);
  const hedgeQuality =
    spreadRet > 0 && notional > 0
      ? Math.max(0, Math.min(1, 1 - wealthGapUsd / (notional * spreadRet)))
      : 1;
  return {
    ...s,
    nYes,
    contracts: Math.round(Math.abs(nYes)),
    premiumUsd: Math.round(Math.abs(premiumSignedUsd)),
    premiumSignedUsd: Math.round(premiumSignedUsd),
    wealthGapUsd,
    hedgeQuality,
  };
}

export function hedgedWealth(
  notional: number,
  ret: number,
  nYes: number,
  yesCost: number,
  yesHappened: boolean,
): number {
  const base = notional * (1 + ret);
  return yesHappened ? base + nYes * (1 - yesCost) : base - nYes * yesCost;
}

File: structuring/hedge_math.py
"""
structuring/hedge_math.py — two-state variance-minimising hedge (hedgingresearch model).

Wealth if YES resolves:  V₀(1 + r_yes)  + N(1 − c)
Wealth if NO resolves:    V₀(1 + r_no)   − N·c

Full offset: N = V₀(r_no − r_yes) on the market's YES contract (N < 0 ⇒ short YES / long NO).
"""
from __future__ import annotations


def yes_returns_from_legs(
    move_adverse: float,
    hedge_leg: str,
    move_favorable: float | None = None,
) -> tuple[float, float]:
    """Map template legs to (ret_if_yes, ret_if_no) for the market's YES contract."""
    mf = 0.0 if move_favorable is None else move_favorable
    if hedge_leg.upper() == "YES":
        return move_adverse, mf
    return mf, move_adverse


def market_yes_prob(hedge_leg: str, leg_price: float) -> float:
    """Implied P(YES) from the side/price we would buy."""
    return leg_price if hedge_leg.upper() == "YES" else 1.0 - leg_price


def residual_pct(sigma_other: float, ret_if_yes: float, ret_if_no: float) -> float:
    spread = abs(ret_if_no - ret_if_yes)
    if spread <= 0:
        return 1.0
    so = sigma_other
    return so * so / (spread * spread + so * so)


def pays_as_hedge(expected_direction: str, move_adverse: float) -> bool:
    """Long equity: contract must pay when the curated adverse state hits."""
    return expected_direction == "adverse" and move_adverse < 0


def approach_note(
    ret_if_yes: float,
    ret_if_no: float,
    p_yes: float,
    n_yes: float,
) -> str:
    """One-line disclosure for UI (mirrors hedgingresearch notebooks)."""
    action = "buy YES" if n_yes >= 0 else "buy NO (short YES)"
    return (
        f"Two-state lock: N = notional × ({ret_if_no:+.0%} − {ret_if_yes:+.0%}) "
        f"at P(YES)={p_yes:.0%}; {action} equalizes wealth across outcomes."
    )


def size_two_state(
    notional: float,
    ret_if_yes: float,
    ret_if_no: float,
    p_yes: float,
    hedge_leg: str,
    *,
    spread: float = 0.0,
    premium_cap_pct: float | None = None,
) -> dict:
    """Size a binary hedge to lock wealth across the two equity states."""
    p = min(max(p_yes, 0.01), 0.99)
    yes_cost = min(p + spread, 0.999)

    n_full = notional * (ret_if_no - ret_if_yes)
    spread_ret = abs(ret_if_no - ret_if_yes)

    budget = (premium_cap_pct * notional) if premium_cap_pct else float("inf")
    if n_full >= 0:
        n_yes = min(n_full, budget / yes_cost) if yes_cost > 0 else n_full
    else:
        n_yes = n_full

    if n_yes >= 0:
        exec_side = "YES"
        contracts = abs(n_yes)
        entry_price = yes_cost
    else:
        exec_side = "NO"
        contracts = abs(n_yes)
        entry_price = 1.0 - yes_cost

    # Actual cash outlay = contracts bought at the executed side's price
    # (YES fills at yes_cost; NO fills at 1 - yes_cost).
    premium_abs = abs(n_yes) * entry_price
    premium = premium_abs

    w_yes = notional * (1 + ret_if_yes) + n_yes * (1 - yes_cost)
    w_no = notional * (1 + ret_if_no) - n_yes * yes_cost
    wealth_gap = abs(w_no - w_yes)

    worst_ret = min(ret_if_yes, ret_if_no)
    l_spread = notional * spread_ret if spread_ret > 0 else 0.0
    worst_pnl_hedged = min(w_yes, w_no) - notional
    worst_pnl_unhedged = notional * worst_ret
    offset = (
        (worst_pnl_hedged - worst_pnl_unhedged) / abs(worst_pnl_unhedged)
        if worst_pnl_unhedged < 0
        else 1.0
    )
    hedge_quality = max(0.0, min(1.0, 1.0 - wealth_gap / l_spread if l_spread > 0 else 1.0))

    return {
        "retIfYes": ret_if_yes,
        "retIfNo": ret_if_no,
        "pYes": p,
        "yesCost": yes_cost,
        "nYes": n_yes,
        "contracts": round(contracts),
        "execSide": exec_side,
        "entryPrice": round(entry_price, 4),
        "premiumUsd": round(premium_abs, 2),
        "premiumSignedUsd": round(premium, 2),
        "wealthGapUsd": round(wealth_gap, 2),
        "dollarOffset": round(offset, 3),
        "hedgeQuality": round(hedge_quality, 3),
        "expectedLossUsd": round(l_spread, 2),
        "approachNote": approach_note(ret_if_yes, ret_if_no, p, n_yes),
    }
